### PR TITLE
mount: Bind-mount root via userns_call

### PR DESCRIPTION
**Pull request by criupatchwork**
This PR is autogenerated from: https://patchwork.criu.org/series/3860

### CODE REVIEW

File: criu/mount.c [hunk 1]
```diff
@@ -2014,7 +2017,10 @@ static int fetch_rt_stat(struct mount_info *m, const char *where)
 static int do_simple_mount(struct mount_info *mi, const char *src, const
 			   char *fstype, unsigned long mountflags)
 {
-	return mount(src, mi->mountpoint, fstype, mountflags, mi->options);
+	int ret = mount(src, mi->mountpoint, fstype, mountflags, mi->options);
+	if (ret)
+		pr_perror("mount(%s, %s)", src, mi->mountpoint);
```
Comment (Snorch) [inline criu/mount.c]: Can you please add mi->mnt_id as mountpoints can be the same for different mounts.

File: criu/mount.c [hunk 2]
```diff
@@ -2501,8 +2507,11 @@ static int do_mount_one(struct mount_info *mi)
 		}
 
 		/* do_mount_root() is called from populate_mnt_ns() */
-		if (mount(opts.root, mi->mountpoint, NULL, MS_BIND | MS_REC, NULL))
+		if (mount(opts.root, mi->mountpoint, NULL, MS_BIND | MS_REC, NULL)) {
+			pr_perror("mount %s %s", opts.root, mi->mountpoint);
```
Comment (Snorch) [inline criu/mount.c]: Also can print mnt_id here.

File: criu/mount.c [hunk 3]
```diff
@@ -2501,15 +2510,33 @@ static int do_mount_one(struct mount_info *mi)
 	pr_debug("\tMounting %s @%s (%d)\n", mi->fstype->name, mi->mountpoint, mi->need_plugin);
 
 	if (rst_mnt_is_root(mi)) {
+		int fd;
+		unsigned long flags = MS_BIND | MS_REC;
```
Comment (Snorch) [inline criu/mount.c]: It's quiet unrelated to the patch, but maybe we can get rid of MS_REC here?
See more info in https://github.com/checkpoint-restore/criu/issues/754

File: criu/cr-restore.c [hunk 4]
```diff
@@ -1589,13 +1589,14 @@ static void restore_pgid(void)
...
-	char proc_mountpoint[] = "crtools-proc.XXXXXX";
+	char proc_mountpoint[] = "/tmp/crtools-proc.XXXXXX";
 
 	if (root_ns_mask == 0)
 		fd = ret = open("/proc", O_DIRECTORY);
 	else {
 		if (mkdtemp(proc_mountpoint) == NULL) {
-			pr_perror("mkdtemp failed %s", proc_mountpoint);
+			pr_perror("Failed to create temp directory %s/%s",
+				get_current_dir_name(), proc_mountpoint);
```
Comment (Snorch) [inline criu/cr-restore.c]: New error message looks inconsistent with the fact that proc_mountpoint is now an absolute path.

**Comment by Snorch** [comment]
And do we need a test for it?

**Comment by avagin** [comment]
@rst0git what do you think about comments to this PR?

**Comment by rst0git** [comment]
I agree with @Snorch. I will update the patch series and add a test case.

File: criu/mount.c (hunk 1 above)
Comment (rst0git) [reply criu/mount.c]: Done

File: criu/mount.c (hunk 2 above)
Comment (rst0git) [reply criu/mount.c]: Done.

File: criu/mount.c (hunk 3 above)
Comment (rst0git) [reply criu/mount.c]: I think this change should be in a separate commit but I'm not sure what would be the result of removing MS_REC. Is it likely to break something?

File: criu/cr-restore.c (hunk 4 above)
Comment (rst0git) [reply criu/cr-restore.c]: Done

**Comment by rst0git** [comment]
I plan to add a test for this change by extending the existing test case for podman after https://github.com/containers/libpod/pull/4525 is merged.

File: criu/mount.c (hunk 3 above)
Comment (Snorch) [reply criu/mount.c]: I don't know anybody who actually uses opts.root on mount with children mounts, and it looks weird for me to do so. Just want to attract some attention to these strange MS_REC flag, it looks like a leftover for me from the past dark times =) But yes it's not connected to your patch directly.

**Comment by avagin** [comment]
@rst0git This test should be in zdtm.py.

Here is my quick draft version.

```
diff --git a/test/zdtm.py b/test/zdtm.py
index 25328104e..36aa30a2a 100755
--- a/test/zdtm.py
+++ b/test/zdtm.py
@@ -62,6 +62,7 @@ tests_root = None
 def clean_tests_root():
     global tests_root
     if tests_root and tests_root[0] == os.getpid():
+        os.rmdir(os.path.join(tests_root[1], "root"))
         os.rmdir(tests_root[1])
 
 
@@ -70,7 +71,9 @@ def make_tests_root():
     if not tests_root:
         tests_root = (os.getpid(), tempfile.mkdtemp("", "criu-root-", "/tmp"))
         atexit.register(clean_tests_root)
-    return tests_root[1]
+        os.chmod(tests_root[1], 0o777)
+        os.mkdir(os.path.join(tests_root[1], "root"))
+    return os.path.join(tests_root[1], "root")
 
 
 # Report generation
@@ -482,6 +485,8 @@ class zdtm_test:
             # Wait less than a second to give the test chance to
             # move into some semi-random state
             time.sleep(random.random())
+        if self.__flavor.ns:
+            os.chmod(os.path.dirname(self.__flavor.root), 0o700)
 
     def kill(self, sig=signal.SIGKILL):
         self.__freezer.thaw()
```

**Pull request closed without merging**